cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -0,0 +1,604 @@
+
+	machineSetResource, _ := schema.ParseResourceArg(fmt.Sprintf("machinesets.%v", defaultMachineAPI))
+	if machineSetResource == nil {
+		panic("MachineSetResource")
+	}
+
+	machineResource, _ := schema.ParseResourceArg(fmt.Sprintf("machines.%v", defaultMachineAPI))
+	if machineResource == nil {
+		panic("machineResource")
+	}
+	machineInformer := informerFactory.ForResource(*machineResource)
+	machineSetInformer := informerFactory.ForResource(*machineSetResource)
+	var machineDeploymentInformer informers.GenericInformer
+
+	machineDeploymentInformer = informerFactory.ForResource(*machineDeploymentResource)
+	machineDeploymentInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{})
+
+	machineInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{})
+	machineSetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{})
+
+	nodeInformer := kubeInformerFactory.Core().V1().Nodes().Informer()
+	nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{})
+
+	if err := machineInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
+		machineProviderIDIndex: indexMachineByProviderID,
+	}); err != nil {
+		return nil, fmt.Errorf("cannot add machine indexer: %v", err)
+	}
+
+	if err := nodeInformer.GetIndexer().AddIndexers(cache.Indexers{
+		nodeProviderIDIndex: indexNodeByProviderID,
+	}); err != nil {
+		return nil, fmt.Errorf("cannot add node indexer: %v", err)
+	}
+
+	return &machineController{
+		kubeInformerFactory:       kubeInformerFactory,
+		machineInformerFactory:    informerFactory,
+		machineDeploymentInformer: machineDeploymentInformer,
+		machineInformer:           machineInformer,
+		machineSetInformer:        machineSetInformer,
+		nodeInformer:              nodeInformer,
+		dynamicclient:             dynamicclient,
+		machineSetResource:        machineSetResource,
+		machineResource:           machineResource,
+		machineDeploymentResource: machineDeploymentResource,
+	}, nil
+}
+
+func (c *machineController) machineSetProviderIDs(machineSet *MachineSet) ([]string, error) {
+	machines, err := c.machinesInMachineSet(machineSet)
+	if err != nil {
+		return nil, fmt.Errorf("error listing machines: %v", err)
+	}
+
+	var providerIDs []string
+	for _, machine := range machines {
+		if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+			klog.Warningf("Machine %q has no providerID", machine.Name)
+		}
+
+		if machine.Spec.ProviderID != nil && *machine.Spec.ProviderID != "" {
+			providerIDs = append(providerIDs, *machine.Spec.ProviderID)
+			continue
+		}
+
+		if machine.Status.NodeRef == nil {
+			klog.V(4).Infof("Status.NodeRef of machine %q is currently nil", machine.Name)
+			continue
+		}
+
+		if machine.Status.NodeRef.Kind != "Node" {
+			klog.Errorf("Status.NodeRef of machine %q does not reference a node (rather %q)", machine.Name, machine.Status.NodeRef.Kind)
+			continue
+		}
+
+		node, err := c.findNodeByNodeName(machine.Status.NodeRef.Name)
```

Comment:
I'm wondering if this could be replaced by finding the node by providerID instead? We require this for setting the NodeRef in ClusterAPI anyway. We do a bit of manipulation of the providerID for comparsion to avoid potential issues with a cloud provider setting additional information (such as region): https://github.com/kubernetes-sigs/cluster-api/blob/6a8ed7875e3d18979f0dd2df230f78cac23a4d8b/controllers/noderefutil/providerid.go#L47

I believe that would also remove the need for interacting with the NodeRef field in an unstructured manner as well..

Reply:
I need to dig up some historical references; pretty sure I looked at this in the past but not everything was supplying providerID. Equally, we could say, that this integration **requires** this to be set.

Reply:
It's a requirement for Cluster API infrastructure providers starting with v1alpha2: https://cluster-api.sigs.k8s.io/architecture/controllers/machine.html?highlight=providerid#infrastructure-provider

Reply:
And OpenStack was another - https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/274

Reply:
> It's a requirement for Cluster API infrastructure providers starting with v1alpha2

OK, good to know. :)

Reply:
Dropping nodeRef would be nice. It simplifies the implementation and expectations for consumers. I guess one decision is: should we drop it in this PR, or as a follow-up. I strongly prefer the latter given the exposure and testing it has had with it being present.

Reply:
I'm good with keeping it as is for the purpose of progress and addressing it later :)